text/0000-semver-for-ts.md

```diff
@@ -0,0 +1,870 @@
+[tsd]: https://github.com/SamVerschueren/tsd
+[tsd-versions]: https://github.com/SamVerschueren/tsd/issues/47
+[expect-type]: https://github.com/mmkal/ts/tree/master/packages/expect-type#readme
+
+In addition to *writing* these tests, package authors should make sure to run the tests (as appropriate to the testing tool chosen) in their continuous integration configuration, so that any changes made to the library are validated to make sure the API has not been changed accidentally.
+
+Further, just as packages are encouraged to test against all a matrix of Ember versions which includes the current stable release, the currently active Ember LTS release, and the canary and beta releases, packages should test the types against all versions of TypeScript supported by the package (see the [suggested policy for version support](#policy-for-supported-typescript-versions) below) as well as the upcoming version (the `next` tag for the `typescript` package on npm).
+
+Type tests can run as normal [ember-try] variations or similar CI. Typed Ember will document a conventional setup for ember-try configurations, so that correct integration into CI setups will be straightforward for package authors.
+
+[ember-try]: https://github.com/ember-cli/ember-try
+
+
+#### Mitigate breaking changes
+
+It is insufficient merely to be *aware* of breaking changes. It is also important to *mitigate* them, to minimize churn and breakage for package users.
+
+
+##### Avoiding user constructability
+
+For types where it is useful to publish an interface for end users, but where users should not construct the interface themselves, authors have a number of options (noting that this list is not exhaustive):
+
+-   The type can simply be documented as non-user-constructable. This is the easiest, and allows an escape hatch for scenarios like testing, where users will recognize that if the public interface changes, they will necessarily need to update their test mocks to match. This can further be mitigated by providing a sanctioned test helper to construct test versions of the types.
+
+-   Export a nominal-like version of the type, using `export type` with a class with a private field:
+
+    <details>
+
+    <summary>implementation of a nominal-like class in TS</summary>
+
+    ```ts
+    class Person {
+      // 1.  The private brand means this cannot be constructed other than the
+      //     class's own constructor, because other approaches cannot add the
+      //     private field. Even if you write a class yourself with a matching
+      //     private field, TS will treat them as distinct.
+      // 2.  Using `declare` means this marker has no runtime over head: it will
+      //     not be emitted by TypeScript or Babel.
+      // 3.  Because the class itself is declared but not exported, the only way
+      //     to construct it is using the function exported lower in the module.
+      declare private __brand: '_Person';
```

Comment:
Maybe this just muddies the water, but this could equally be `brand: void` or any other type and have the same effect.

Reply:
Nah, I think that's useful to make the point. Updated in 033c142